src/Features/Core/Portable/DesignerAttributes/AbstractDesignerAttributeService.cs

```diff
@@ -63,75 +68,76 @@ public async Task<DesignerAttributeResult> ScanDesignerAttributesAsync(Document
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    // if it has designer attribute, set it
+                    var attribute = type.GetAttributes().Where(d => designerAttribute.Equals(d.AttributeClass)).FirstOrDefault();
+                    if (attribute != null && attribute.ConstructorArguments.Length == 1)
                     {
-                        if (type.IsErrorType())
-                        {
-                            documentHasError = true;
-                            continue;
-                        }
-
-                        cancellationToken.ThrowIfCancellationRequested();
-
-                        // if it has designer attribute, set it
-                        var attribute = type.GetAttributes().Where(d => designerAttribute.Equals(d.AttributeClass)).FirstOrDefault();
-                        if (attribute != null && attribute.ConstructorArguments.Length == 1)
-                        {
-                            designerAttributeArgument = GetArgumentString(attribute.ConstructorArguments[0]);
-                            return new DesignerAttributeResult(designerAttributeArgument, documentHasError, applicable: true);
-                        }
+                        var designerAttributeArgument = GetArgumentString(attribute.ConstructorArguments[0]);
+                        return new DesignerAttributeResult(designerAttributeArgument, documentHasError, applicable: true);
                     }
                 }
+            }
+
+            return new DesignerAttributeResult(
+                designerAttributeArgument: null, documentHasError, applicable: true);
+        }
+
+        private TClassDeclarationSyntax GetFirstTopLevelClass(ISyntaxFactsService syntaxFacts, TCompilationUnitSyntax compilationUnit)
+            => GetFirstTopLevelClass(syntaxFacts, syntaxFacts.GetMembersOfCompilationUnit(compilationUnit));
 
-                // check only first type
-                if (ProcessOnlyFirstTypeDefined())
+        private TClassDeclarationSyntax GetFirstTopLevelClass(
+            ISyntaxFactsService syntaxFacts, SyntaxList<SyntaxNode> members)
+        {
+            foreach (var member in members)
+            {
+                if (member is TNamespaceDeclarationSyntax namespaceDeclaration)
+                {
+                    var classNode = GetFirstTopLevelClass(
+                        syntaxFacts, syntaxFacts.GetMembersOfNamespaceDeclaration(namespaceDeclaration));
+                    if (classNode != null)
+                        return classNode;
+                }
+                else if (member is TClassDeclarationSyntax classDeclaration)
                 {
-                    break;
+                    return classDeclaration;
                 }
             }
 
-            return new DesignerAttributeResult(designerAttributeArgument, documentHasError, applicable: true);
+            return null;
```

Comment:
I don't think sharing this code is worth the extra complexity. 

I think it'd be better to have a single abstract method 
```C#
SyntaxNode FindPotentialDesignerCategoryAttributeTarget(Document document);
```

that encapsulates all the language specific syntactic details.

Reply:
this is the pattern we use across teh IDE when there are no differences in the syntactic details (except that there different syntax classes).

if we write this for Vb and C# it just means we ahve to keep the logic manually in sync.

Reply:
also, much of this will likely go away with the OOP work.  but this was a simple simplification i noticed while wrapping my head aronud what the current code did.  by simplifying (and pushing that into another PR), it just made the OOP side simpler (since i could then more easily compare the new OOP logic to the changed logic here instead of the original much more complex logic).

Reply:
> we have to keep the logic manually in sync.

I think that's fine. The logic is trivial. I'd agree if the logic was complex, but that's not the case here.